### PR TITLE
Join stake pool endpoint (part 1 - no signCert impl)

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#896 

# Overview

- [x] I have added pools check in endpoint
- [x] I have added createCert
- [x] I have added submitCert 
- [x] I have added computeCertFee 
- [x] I have added test and adapted other tests affected 

# Comments

In the forthcoming PR I will tackle signCert impl (waiting here for mkDelegationCertTx to find its way into master first)

### CODE REVIEW

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet.hs]: waiting for `mkCertificateTx` PRs to be merged

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: I need to think once again if submitTx can be used here, especially in the context of its interaction to DB

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet.hs]: writing both tx history and delegation certificate in DB

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: changing wallet metadata. now delegating with pool id. wonder if we need to check first if wallet is already not delegating. if yes, then error is thrown as we need to first quickStakePool. I have heard two versions, this one. and also that joinStakePool just changes this field. which means changes the stake pool it delegates. Another question. what to do if we are already joined to the pool and want to rejoin to the same one. I am tentatively in favor of also throwing error. Any thought/suggestion here?

File: lib/core/src/Cardano/Wallet/Primitive/Fee.hs
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/Fee.hs]: I am here assuming that this is flat fee for certificates and I can delegate to one stake pool. If this assumption is not true, then I should change the signature of this function and add `number of certificates`. On the other hand I see, something like this : https://slack-redir.net/link?url=https%3A%2F%2Fgithub.com%2Finput-output-hk%2Fchain-libs%2Fblob%2Fmaster%2Fchain-impl-mockchain%2Fdoc%2Fformat.md%23type-0-initial-blockchain-configuration and here 
```
LinearFee has the format:

Constant | Coefficient | Certificate
all of them 64-bit unsigned integers, specifying how fees are computed using the formula:

Constant + Coefficient * (inputs + outputs) + Certificate * certificates
where inputs, outputs and certificates represent the size of the serialization of the corresponding parts of a transaction in bytes.
```
so according to this document `certificates` is size of serialization part

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (paweljakubas) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: I know that I need to use `pIdAbsent` here in the end. I just get here 
```
\"4f8d686a02c6e625b5a59cc9e234f32e5d72987012f9c25c9a6b60ddade197d9\"
```
when using `show pIdAbsent`

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: :+1:

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: style sugggestion: 

```hs
signCert ctx wid argGenChange pwd coinSel _poolId = db & \DBLayer{..} -> do
    let (CoinSelection ins outs chgs) = coinSel
    ...
```

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: The shuffling is so-to-speak pointless in this scenario because all outputs are already ours. In the case of standard transaction, we shuffle to add some privacy and avoid having outputs that follow a very specific pattern:

- recipient
- change
- recipient
- change
etc ...

Here, there's no need :)

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: Generally speaking, I'd suggest putting an explicit `error "not implemented"` and avoid writing functions that:

1. Compiles
2. Raises no runtime exception
3. Are not correct

In this case, I am pretty sure that we won't forgot to update it, but, as a habit, it's good to make sure that "if it compiles, it works".  If it doesn't work, either make sure it doesn't compile, or make sure it fails loudly :)

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: :+1: I see room for code factorization with the `signTx` function that is very analogous. I wouldn't recommend doing it in this PR, but definitely leave a bullet point in the development plan of the ticket about refactoring this.

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet.hs]: :ok_hand: 

suggestion:

```hs
putDelegationCertificate (PrimaryKey wid) poolId (meta ^. #slotId)
```

why ? Avoiding pattern-maching here makes for code that is slightly easier to maintain. Adding fields to `TxMeta` won't require us to go back to this very point to add an extra `_`. Alternatively, using `NamedFieldPuns`, I'd also favor the following approach:

```hs
let TxMeta{sl} = meta
putDelegationCertificate (PrimaryKey wid) poolId sl
```

I'd still favor the first one because the entire line is then captured in a single expression and is easier to load in brain :)

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Ah! But this is actually slightly _wrong_. We haven't joined the stake pool at the very stage and aren't yet delegating. 
We have only _made a request_ about it. The delegation will only be effective in the epoch following the insertion of the certificate in the ledger. Detecting that we are delegating is a job for #899 ! So here, there's no need to update the wallet metadata, they'll be updated when the certificate is discovered.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: Can't we re-use previous handlers here?

File: lib/core/src/Cardano/Wallet/Primitive/Fee.hs
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/Fee.hs]: I agree that this function should probably take an extra argument to reflect the number of certificates. And, at the call-site, use a value of `1` since this is where we actually _know_ that we are only creating a single certificate.

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (KtorZ) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: Alternatively, use `0000000.....00000` for an unknown stake pool. Chances of getting a stake pool with such id are null (1 in 2^32 exactly)

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs]: opted for lens version

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: thanks for this remark. removed this. hope we will tackle this in #899

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (paweljakubas) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: used 00000....00000 and refactored the test case a little bit

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs]: done

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs]: indeed, corrected

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs]: my intention was to wait for mkCertificateTx to be merged, then rebase this PR and only when the functionalities work as we expect and all integration tests pass, merge this PR (according to the rule, that only well tested changes can be merged, no partial, untested code in master etc)

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs]: added a point in development plan : 
```
[ ] Consider reusing create/sign/submit parts between Tx and Cert
```

**Comment by KtorZ** [comment]
@paweljakubas do you think of any way to split that PRs into smaller ones that aren't dependent on other PRs? 

In particular, I think that the whole API part could be merged already if we replaced the wallet engine function implementations by `error "createCert: not implemented"`. 

What do you think?

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: **[change required]**
This line should be removed. 

**Why**
We don't keep track of "pending delegation certificates" the same way we keep track of transaction. So, until a certificate made it into the ledger (and is therefore, discover by the restoration loop), it simply doesn't exist. In particular, the "slotId" would be quite wrong here; it should refer to the slot at which the certificate is inserted, not the one at which we made the request (which has very little meaning from the blockchain PoV).

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: **[suggestion: readability]**

I'd suggest to move the guard in a dedicated function (possibly declared in the where clause) to have something which is more consistent with the style adopted so far:

```hs
whenPoolIsUnknown poolId  $ liftHandler $ throwE (ErrCreateCertNoSuchPool poolId)
```

**Why**

This keeps the `joinStakePool` function more readable by moving some lower-level concerns into a separate function. Then, `joinStakePool` reads more like a sentence.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: **[suggestion: potentially error prone]**
Remove this line.

**Why**
`signCert` will already throw loudly and cause the request to fail with a 500. So this looks redundant to me. Otherwise, we need to remember to modify this section of the code after implementing `signCert`. This could be confusing, although, this is very minor.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: **[change required]**
This should rather be `err403`

**Why**
`404` is used for missing pages or resources. Here, the case is slightly different: the pool exists, the request is syntactically valid, but we refuse to proceed. That's the semantic given to `403 Forbidden`

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (KtorZ) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:163]: :+1:

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (KtorZ) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:171]: **[change required]**
`joinStakePool` requires a full `ApiStakePool` as argument which forces to create dummy values. I'd suggest to either:

1. drop `joinStakePool` and simply use the `request` combinator from the integration framework. 
2. Review `joinStakePool` and `joinStakePoolEp` to only require a stake pool id instead of a whole body

**Why**
The goal of the integration DSL is to favor readability and strive to have readable integration scenarios. By adding extra non-orthogonal (a.k.a overlapping) combinators, we do actually create noise and reduce readability and maintenance in the long run. Here, we are forced to create a fake pool metrics and a fake `ApiStakePool` object, whereas, we really just need to write:

```hs
    let poolId = T.replicate 64 "0"
    w <- emptyWallet ctx
    let payload = Json [aesonQQ| { "passphrase": "Secure Passphase" } |]
    request @(ApiTransaction 'Testnet) ctx (joinStakePoolEp poolId w) Default payload
```

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs]: done

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: done

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: indeed, done

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: corrected

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (paweljakubas) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:171]: I have spotted this, and planned to review it in forthcoming PR (here just used the same way as in other examples in the test suite). Added a point in development plan : 
```
[ ] review joinStakePoolEp in DSL
```
to tackle this in #896

**Comment by KtorZ** [comment]
_doing a 3rd round of review_

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs:1041]: :+1:

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:833]: :+1:

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:826]: :+1:

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:1646]: :+1:

**Review by KtorZ — APPROVED** [review]
LGTM.

Perhaps a bit of cleanup / housekeeping the commit history before merging.

File: lib/core/src/Cardano/Wallet.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs:118]: Same point as I made here https://github.com/input-output-hk/cardano-wallet/pull/1059#discussion_r349318171: 

Because there are multiple kinds of certificates in jormungandr I find these names a bit ambiguous.

What do you think about:
- `createDelegationCert`
- `signDelegationCert`
- `submitDelegationCert`
- `ErrCreateDelegationCert`

File: lib/core/src/Cardano/Wallet.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs:1002]: 👌

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:824]: I'm not sure if this is wise.
1. `listStakePools` is not a reliable way to list the current `[PoolId]` because it has to be in sync with its chain-follower.*
2. This is using the api-`listPools`. So the error-messages from that endpoint can be thrown here I presume. I had never realised it was possible to nest api-calls like this though — cool! — but I don't think we want it?

Could we use `NetworkLayer.stakeDistribution` instead?

**Review by Anviking — APPROVED** [review]
lgtm! **But**, I don't think we'd like to use `listPools` from `joinStakePool`.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:824]: oh, thanks for this remark. As this is not the only one PR in this task, I will address it in the signCert (or signDelegationCert) PR soon

File: lib/core/src/Cardano/Wallet.hs
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs:118]: sound argument, I will address it in the signCert (or signDelegationCert) PR soon

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet.hs:118]: I agree that naming could be clearer but in the context of the wallet, the term certificate is non-ambiguous. Jörmungandr is out of the question at this level (as well as in the API). 

Yet, it wouldn't hurt to add `Delegation` to make things clearer indeed.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:824]: what about simply relying on `DB.readPoolProduction` instead to avoid unnecessary networks roundtrips?

Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:824]: > what about simply relying on DB.readPoolProduction instead to avoid unnecessary networks roundtrips?

In order for the `DB` to contain any stake-distributions (would be `DB.readStakeDistribution`), it does at least need to _begin_ following the chain correctly. In contrast to the wallet, it needs to (I think it was)`fetchTip >> fetchStakeDistribution >> fetchTip` without the tip changing, which is a point of failure  unnecessary to bring in here.

Doing that would entangle the stake-pool & wallet state in a way I am worried about.

The network roundtrip should be _really_ insignificant here, anyway?

**Pull request merged**
